t/t5799-gvfs-helper.sh

```diff
@@ -0,0 +1,972 @@
+#!/bin/sh
+
+test_description='test gvfs-helper and GVFS Protocol'
+
+. ./test-lib.sh
+
+# Set the port for t/helper/test-gvfs-protocol.exe from either the
+# environment or from the test number of this shell script.
+#
+test_set_port GIT_TEST_GVFS_PROTOCOL_PORT
+
+# Setup the following repos:
+#
+#    repo_src:
+#        A normal, no-magic, fully-populated clone of something.
+#        No GVFS (aka VFS4G).  No Scalar.  No partial-clone.
+#        This will be used by "t/helper/test-gvfs-protocol.exe"
+#        to serve objects.
+#
+#    repo_t1:
+#        An empty repo with no contents nor commits.  That is,
+#        everything is missing.  For the tests based on this repo,
+#        we don't care why it is missing objects (or if we could
+#        actually use it).  We are only testing explicit object
+#        fetching using gvfs-helper.exe in isolation.
+#
+REPO_SRC="$PWD"/repo_src
+REPO_T1="$PWD"/repo_t1
+
+# Setup some loopback URLs where test-gvfs-protocol.exe will be
+# listening.  We will spawn it directly inside the repo_src directory,
+# so we don't need any of the directory mapping or configuration
+# machinery found in "git-daemon.exe" or "git-http-backend.exe".
+#
+# This lets us use the "uri-base" part of the URL (prior to the REST
+# API "/gvfs/<token>") to control how our mock server responds.  For
+# example, only the origin (main Git) server supports "/gvfs/config".
+#
+# For example, this means that if we add a remote containing $ORIGIN_URL,
+# it will work with gvfs-helper, but not for fetch (without some mapping
+# tricks).
+#
+HOST_PORT=127.0.0.1:$GIT_TEST_GVFS_PROTOCOL_PORT
+ORIGIN_URL=http://$HOST_PORT/servertype/origin
+CACHE_URL=http://$HOST_PORT/servertype/cache
+
+SHARED_CACHE_T1="$PWD"/shared_cache_t1
+
+# The pid-file is created by test-gvfs-protocol.exe when it starts.
+# The server will shut down if/when we delete it.  (This is a little
+# easier than killing it by PID.)
+#
+PID_FILE="$PWD"/pid-file.pid
+SERVER_LOG="$PWD"/OUT.server.log
+
+PATH="$GIT_BUILD_DIR/t/helper/:$PATH" && export PATH
+
+OIDS_FILE="$PWD"/oid_list.txt
+OIDS_CT_FILE="$PWD"/oid_ct_list.txt
+OIDS_BLOBS_FILE="$PWD"/oids_blobs_file.txt
+OID_ONE_BLOB_FILE="$PWD"/oid_one_blob_file.txt
+
+# Get a list of available OIDs in repo_src so that we can try to fetch
+# them and so that we don't have to hard-code a list of known OIDs.
+# This doesn't need to be a complete list -- just enough to drive some
+# representative tests.
+#
+# Optionally require that we find a minimum number of OIDs.
+#
+get_list_of_oids () {
+	git -C "$REPO_SRC" rev-list --objects HEAD | sed 's/ .*//' | sort >"$OIDS_FILE"
+
+	if test $# -eq 1
+	then
+		actual_nr=$(( $(wc -l <"$OIDS_FILE") ))
```

Comment:
The idiomatic way to do this in Git's test suite is `actual_nr=$(wc -l "$OIDS_FILE")`.

Reply:
I had several problems getting this function to work as I wanted and settled on this.
Let me ask a few questions about the problems that I fought thru to get this working.

When I do `x=$(wc -l file)`, I get the count and the filename in the output.

When I do `x=$(wc -l <file)`, I just get the count in the output.  This saves me from having to run sed/awk on the output, right?

On the Mac, `wc -l` yields a string with leading spaces.  This doesn't play right with the `-lt` or `-eq` operators.  Wrapping the expression in `$(( ... ))` gives me a number which behaved correctly.

Does the variable assignment eat the leading whitespace, such that `x=$(wc -l <file)` is the right answer ??

Reply:
> When I do `x=$(wc -l file)`, I get the count and the filename in the output.

I forgot about that, sorry.

> When I do `x=$(wc -l <file)`, I just get the count in the output. This saves me from having to run sed/awk on the output, right?

Yes.

> On the Mac, `wc -l` yields a string with leading spaces. This doesn't play right with the `-lt` or `-eq` operators. Wrapping the expression in `$(( ... ))` gives me a number which behaved correctly.

Right, but if you do not wrap `$actual_nr` in double-quotes, the string is white-space trimmed implictly, too.

> Does the variable assignment eat the leading whitespace, such that `x=$(wc -l <file)` is the right answer ??

Not quite. The quoting is not necessary here, but I think it will still reduce the whitespace to a single space on macOS. Leaving off the double-quotes when expanding `$actual_nr` is the actual trick.